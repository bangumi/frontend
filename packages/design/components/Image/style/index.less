@loadingColor: #EDE9E9;
@outlineColor: rgba(255, 255, 255, 0.4);
@boxShadowColor: rgba(237, 233, 233, 0.8);

.bgm-image {
  display: inline-block;
  overflow: hidden;
  background: @loadingColor;

  &>img {
    object-fit: cover;
  }

  &:hover {
    outline: 2px solid @outlineColor;
    outline-offset: -2px;
  }

  &--withBoxShadow {
    box-shadow: 0px 5px 10px 3px @boxShadowColor;
  }

  &--rect {
    border-radius: 6px;
  }

  &--circle {
    border-radius: 50%;
  }
}