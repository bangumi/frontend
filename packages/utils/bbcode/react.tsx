import React from 'react';

import { convert } from './convert';
import { Parser } from './parser';
import type { CodeNodeTypes, NodeTypes, VNode } from './types';

/* eslint-disable no-useless-escape */
const toCamelCase = (s: string) => s.replace(/(\-\w)/g, (k) => k[1]?.toUpperCase() ?? '');

const joinClassName = (cls: string | string[]) => {
  if (Array.isArray(cls)) {
    return cls.join(' ');
  }
  return cls;
};

const convertToReactProps = (node: VNode, key?: React.Key) => {
  const { props, style, className } = node;

  const transformedStyle: Record<string, string> = style
    ? Object.entries(style).reduce((pre, [key, val]) => {
        return {
          ...pre,
          [toCamelCase(key)]: val,
        };
      }, {})
    : {};

  const transformedClassName = className !== undefined ? joinClassName(className) : undefined;

  const reactProps = {
    ...props,
    style: transformedStyle,
    className: transformedClassName,
    key,
  };

  return reactProps;
};

export const renderNode = (node: NodeTypes, key?: React.Key): React.ReactElement | string => {
  if (typeof node === 'string') {
    const splittedStr: Array<string | React.ReactElement> = node.split(/\n/g);

    if (splittedStr.length < 2) {
      return splittedStr[0] ?? '';
    }
    const children = splittedStr.reduce<Array<string | React.ReactElement>>((pre, curr, idx) => {
      if (idx > 0) {
        return [...pre, React.createElement('br', null), curr];
      }
      return [...pre, curr];
    }, []);
    return React.createElement(React.Fragment, { key }, ...children);
  }

  const { type, children } = node;

  if (children) {
    return React.createElement(
      type,
      convertToReactProps(node, key),
      ...children.map((child, idx) => renderNode(child, idx)),
    );
  }
  return React.createElement(type, convertToReactProps(node, key));
};

export const render = (rawStr: string): Array<React.ReactElement | string> => {
  const nodes: CodeNodeTypes[] = new Parser(rawStr).parse();

  return nodes.map((node, idx) => renderNode(convert(node, {}), idx));
};
