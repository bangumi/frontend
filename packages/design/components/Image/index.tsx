import './style';

import classnames from 'classnames';
import type { FC, ImgHTMLAttributes } from 'react';
import React from 'react';

export interface ImageProps extends Omit<ImgHTMLAttributes<HTMLImageElement>, 'src' | 'alt'> {
  /** 图像的 URL */
  src: string;
  /** 图像的备用文本描述 */
  alt?: string;
  /** 是否增加 box-shadow */
  withBoxShadow?: boolean;
  /** 形状, 矩形或圆形, 默认为矩形 */
  shape?: 'rect' | 'circle';
}

const Image: FC<ImageProps> = ({
  src,
  width,
  height,
  withBoxShadow,
  className,
  shape = 'rect',
  ...rest
}) => {
  const actualClassnames = classnames('bgm-image', className, {
    'bgm-image--withBoxShadow': withBoxShadow,
    'bgm-image--rect': shape === 'rect',
    'bgm-image--circle': shape === 'circle',
  });
  return <img className={actualClassnames} src={src} data-testid='img-wrapper' {...rest} />;
};

export default Image;
