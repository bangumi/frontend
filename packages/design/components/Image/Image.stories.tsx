import React from 'react'
import { ComponentStory, ComponentMeta } from '@storybook/react'
import Image from '.'

const src = 'data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTM2IiBoZWlnaHQ9IjIwMCIgdmlld0JveD0iMCAwIDEzNiAyMDAiIGZpbGw9Im5vbmUiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+DQo8cmVjdCB3aWR0aD0iMTM2IiBoZWlnaHQ9IjIwMCIgcng9IjYiIGZpbGw9ImFxdWEiLz4NCjwvc3ZnPg0K'

const componentMeta: ComponentMeta<typeof Image> = {
  title: 'modern/Image',
  component: Image
}

export default componentMeta

const Template: ComponentStory<typeof Image> = (args) => {
  return <Image {...args} />
}

export const Loaded = Template.bind({})
Loaded.args = {
  src,
  alt: 'vanilla',
  width: 136,
  height: 200
}

export const Loading = Template.bind({})
Loading.args = {
  src: 'https://www.deelay.me/3500/https://lain.bgm.tv/pic/cover/c/c2/4c/253_t3XWb.jpg',
  alt: 'vanilla',
  width: 150,
  height: 205
}
