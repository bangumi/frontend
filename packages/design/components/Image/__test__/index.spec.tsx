import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import Image from '..';

const src =
  'data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTM2IiBoZWlnaHQ9IjIwMCIgdmlld0JveD0iMCAwIDEzNiAyMDAiIGZpbGw9Im5vbmUiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+DQo8cmVjdCB3aWR0aD0iMTM2IiBoZWlnaHQ9IjIwMCIgcng9IjYiIGZpbGw9ImFxdWEiLz4NCjwvc3ZnPg0K';

describe('<Image />', () => {
  it('should render correctly', () => {
    const { getByTestId } = render(<Image src={src} alt='vanilla' />);

    expect(getByTestId('img-wrapper')).toMatchSnapshot();
  });

  it('should render box-shadow properly', () => {
    const { getByTestId } = render(<Image src={src} alt='vanilla' withBoxShadow />);

    expect(getByTestId('img-wrapper')).toMatchSnapshot();
  });

  it('should render hovering styles properly', () => {
    const { getByTestId } = render(<Image src={src} alt='vanilla' withBoxShadow />);

    const Img = getByTestId('img-wrapper');

    fireEvent.mouseEnter(Img);

    expect(Img).toMatchSnapshot();
  });

  it('should render circle shape properly', () => {
    const { getByTestId } = render(<Image src={src} alt='vanilla' withBoxShadow shape='circle' />);

    expect(getByTestId('img-wrapper')).toMatchSnapshot();
  });
});
