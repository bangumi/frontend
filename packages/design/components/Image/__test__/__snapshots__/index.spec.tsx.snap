// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Image /> > should render box-shadow properly 1`] = `
<img
  alt="vanilla"
  class="bgm-image bgm-image--withBoxShadow bgm-image--rect"
  data-testid="img-wrapper"
  src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTM2IiBoZWlnaHQ9IjIwMCIgdmlld0JveD0iMCAwIDEzNiAyMDAiIGZpbGw9Im5vbmUiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+DQo8cmVjdCB3aWR0aD0iMTM2IiBoZWlnaHQ9IjIwMCIgcng9IjYiIGZpbGw9ImFxdWEiLz4NCjwvc3ZnPg0K"
/>
`;

exports[`<Image /> > should render circle shape properly 1`] = `
<img
  alt="vanilla"
  class="bgm-image bgm-image--withBoxShadow bgm-image--circle"
  data-testid="img-wrapper"
  src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTM2IiBoZWlnaHQ9IjIwMCIgdmlld0JveD0iMCAwIDEzNiAyMDAiIGZpbGw9Im5vbmUiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+DQo8cmVjdCB3aWR0aD0iMTM2IiBoZWlnaHQ9IjIwMCIgcng9IjYiIGZpbGw9ImFxdWEiLz4NCjwvc3ZnPg0K"
/>
`;

exports[`<Image /> > should render correctly 1`] = `
<img
  alt="vanilla"
  class="bgm-image bgm-image--rect"
  data-testid="img-wrapper"
  src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTM2IiBoZWlnaHQ9IjIwMCIgdmlld0JveD0iMCAwIDEzNiAyMDAiIGZpbGw9Im5vbmUiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+DQo8cmVjdCB3aWR0aD0iMTM2IiBoZWlnaHQ9IjIwMCIgcng9IjYiIGZpbGw9ImFxdWEiLz4NCjwvc3ZnPg0K"
/>
`;

exports[`<Image /> > should render hovering styles properly 1`] = `
<img
  alt="vanilla"
  class="bgm-image bgm-image--withBoxShadow bgm-image--rect"
  data-testid="img-wrapper"
  src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTM2IiBoZWlnaHQ9IjIwMCIgdmlld0JveD0iMCAwIDEzNiAyMDAiIGZpbGw9Im5vbmUiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+DQo8cmVjdCB3aWR0aD0iMTM2IiBoZWlnaHQ9IjIwMCIgcng9IjYiIGZpbGw9ImFxdWEiLz4NCjwvc3ZnPg0K"
/>
`;
